hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bucket/ConsistentBucketStreamWriteFunction.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.hudi.configuration.FlinkOptions;
+import org.apache.hudi.index.bucket.ConsistentBucketIdentifier;
+import org.apache.hudi.index.bucket.ConsistentBucketIndexUtils;
+import org.apache.hudi.sink.StreamWriteFunction;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.streaming.api.functions.ProcessFunction;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A stream write function with consistent hashing bucket index.
+ *
+ * @param <I> the input type
+ */
+public class ConsistentBucketStreamWriteFunction<I> extends StreamWriteFunction<I> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ConsistentBucketStreamWriteFunction.class);
+
+  private List<String> indexKeyFields;
+  private Map<String, ConsistentBucketIdentifier> partitionToIdentifier;
+  /**
+   * Cache mapping between node -> location, to avoid repeated creation of the same location objects.
+   */
+  private Map<ConsistentHashingNode, HoodieRecordLocation> nodeToRecordLocation;
+
+  /**
+   * Constructs a StreamingSinkFunction.
+   *
+   * @param config The config options
+   */
+  public ConsistentBucketStreamWriteFunction(Configuration config) {
+    super(config);
+  }
+
+  @Override
+  public void open(Configuration parameters) throws IOException {
+    super.open(parameters);
+    this.indexKeyFields = Arrays.asList(config.getString(FlinkOptions.INDEX_KEY_FIELD).split(","));
+
+    this.partitionToIdentifier = new HashMap<>();
+    this.nodeToRecordLocation = new HashMap<>();
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    super.initializeState(context);
+  }
+
+  @Override
+  public void snapshotState() {
+    super.snapshotState();
+  }
+
+  @Override
+  public void processElement(I value, ProcessFunction<I, Object>.Context ctx, Collector<Object> out) throws Exception {
+    HoodieRecord<?> record = (HoodieRecord<?>) value;
+    final HoodieKey hoodieKey = record.getKey();
+    final String partition = hoodieKey.getPartitionPath();
+
+    ConsistentHashingNode node = getBucketIdentifier(partition).getBucket(hoodieKey, indexKeyFields);
+    Preconditions.checkArgument(!StringUtils.isNullOrEmpty(node.getFileIdPrefix()),
+        "Consistent hashing node has no file group, partition: " + partition + ", meta: "
+            + partitionToIdentifier.get(partition).getMetadata().getFilename() + ", record_key: " + hoodieKey);
+
+    record.unseal();
+    record.setCurrentLocation(nodeToRecordLocation.computeIfAbsent(node,
+        n -> new HoodieRecordLocation("U", FSUtils.createNewFileId(n.getFileIdPrefix(), 0))));
```

Comment:
Maybe should not set to "U" for first time insert

Reply:
I took another look at the code. Differently from SIMPLE bucket index, setting `I` for first write in the context of consistent hashing would require checking the existence of file slices, which introduces a huge overhead. And we may need to handle race conditions when we have concurrent clustering.

Setting it to `U` will produce log-file-only file group for the first write and will not break any constraints regarding convention or correctness. And the benefit is a performance gain since we no longer need to obtain a FileSystemView and do the check. I will stick to `U` for now. What is your opinion?